misc/refseq-historical-backfill/ncbi_extract_gbff.py

```diff
@@ -0,0 +1,196 @@
+from typing import Iterable
+
+from Bio.Seq import Seq
+import Bio.SeqIO
+from Bio.SeqRecord import SeqRecord
+
+from uta.formats.geneaccessions import GeneAccessions, GeneAccessionsWriter
+from uta.formats.geneinfo import GeneInfo, GeneInfoWriter
+from uta.formats.txinfo import TxInfo, TxInfoWriter
+from uta.parsers.seqrecord import SeqRecordFacade, SeqRecordFeatureError
+from uta.tools.file_utils import open_file
+
+
+logging_conf_fn = importlib_resources.files("uta").joinpath("etc/logging.conf")
+logging.config.fileConfig(logging_conf_fn)
+logging.getLogger().setLevel(logging.INFO)
+logger = logging.getLogger(__name__)
+
+
+def parse_args():
+    ap = argparse.ArgumentParser(
+        description=__doc__,
+    )
+    ap.add_argument("GBFF_FILES", nargs="+")
+    ap.add_argument("--origin", "-o", default="NCBI")
+    ap.add_argument("--prefix", "-p", default="")
+    ap.add_argument("--output_dir", "-d", default=".", type=str)
+    opts = ap.parse_args()
+    return opts
+
+
+def main(gbff_files: Iterable, origin: str, prefix: str, output_dir: str) -> None:
+    if prefix:
+        prefix = f"{prefix}."
+
+    # setup context managers for file writers
+    with ExitStack() as stack:
+        # DNA fasta file
+        dna_fasta_fh = stack.enter_context(
+            io.TextIOWrapper(
+                gzip.open(f"{output_dir}/{prefix}rna.fna.gz", "wb"), encoding="utf-8"
+            )
+        )
+
+        # Protein fasta file
+        protein_fasta_fh = stack.enter_context(
+            io.TextIOWrapper(
+                gzip.open(f"{output_dir}/{prefix}protein.faa.gz", "wb"),
+                encoding="utf-8",
+            )
+        )
+
+        geneinfo_fh = stack.enter_context(
+            io.TextIOWrapper(
+                gzip.open(f"{output_dir}/{prefix}geneinfo.gz", "wb"), encoding="utf-8"
+            )
+        )
+        geneinfo_writer = GeneInfoWriter(geneinfo_fh)
+
+        txinfo_fh = stack.enter_context(
+            io.TextIOWrapper(
+                gzip.open(f"{output_dir}/{prefix}txinfo.gz", "w"), encoding="utf-8"
+            )
+        )
+        txinfo_writer = TxInfoWriter(txinfo_fh)
+
+        assocacs_fh = stack.enter_context(
+            io.TextIOWrapper(
+                gzip.open(f"{output_dir}/{prefix}assocacs.gz", "w"), encoding="utf-8"
+            )
+        )
+        assocacs_writer = GeneAccessionsWriter(assocacs_fh)
+
+        total_genes = set()
+        total_skipped = set()
+        all_prefixes = Counter()
+        for gbff_fn in gbff_files:
+            logger.info(f"Processing {gbff_fn}")
+            gbff_file_handler = stack.enter_context(open_file(gbff_fn))
+            i = 0
+            genes = set()
+            skipped = set()
+            prefixes = Counter()
+            for r in Bio.SeqIO.parse(gbff_file_handler, "gb"):
+                srf = SeqRecordFacade(r)
+
+                # skip transcripts where the exon structure is unknown
+                if not srf.exons_se_i:
```

Comment:
This is where we are skipping genbank records that do not have an exon structure defined.